Engine/EditorUI/Panel/PanelStateMachine.cpp

```diff
@@ -89,52 +126,111 @@ void PanelStateMachine::RenderStates()
 		}
-
 		if (node->state->clip)
 		{
 			ImGui::Checkbox("Loop", &node->state->clip->loop);
-			ImGui::SameLine();
+			modified_by_user = true;
+			ImGui::SameLine(node->Size.x/2 + 20);
 		}
 		bool is_default_state = node->state->name_hash == state_machine->default_state;
 		if (ImGui::Checkbox("Default", &is_default_state))
 		{
 			state_machine->default_state = node->state->name_hash;
+			modified_by_user = true;
 		}
-	
+		
+		if (ImGui::InputFloat("Speed", &node->speed, 0.5f, 0.0f, "%.2f", 0))
+		{
+			node->state->speed = node->speed;
+			modified_by_user = true;
+		}
+
 		ImGui::BeginGroup();
 		ax::NodeEditor::BeginPin(node->input, ax::NodeEditor::PinKind::Input);
-		ImGui::Text("-> In");
+		Widgets::Icon(node->Size/5, ax::Drawing::IconType::Diamond, true, ImVec4(255, 127, 80, 255), ImVec4(255, 127, 80, 255));
 		ax::NodeEditor::EndPin();
 		ImGui::EndGroup();
-		ImGui::SameLine();
+
+		ImGui::SameLine(node->Size.x - 1.5);
+
 		ImGui::BeginGroup();
 		ax::NodeEditor::BeginPin(node->output, ax::NodeEditor::PinKind::Output);
-		ImGui::Text("Out ->");
+		Widgets::Icon(node->Size / 5, ax::Drawing::IconType::Flow, true, ImVec4(0, 0, 255, 255), ImVec4(0, 0, 255, 255));
 		ax::NodeEditor::EndPin();
 		ImGui::EndGroup();
+
 		ax::NodeEditor::EndNode();
+		
+		selected_nodes = GetSelectedNodes();
+
 		if (ImGui::IsMouseClicked(1) && ImGui::IsItemHovered())
 		{
 			ax::NodeEditor::Suspend();
 			ImGui::OpenPopup("Create State");
 			ax::NodeEditor::Resume();
 		}
+		//update node->position
 		position = ax::NodeEditor::GetNodePosition(node->id);
 		position.x+= ax::NodeEditor::GetNodeSize(node->id).x;
+		node->Pos = position;
 
 		ImGui::PopItemWidth();
 		ImGui::PopID();
+		
+		//get the current running animation controller to verify playing clip and select corresponding node/link
+		//if (animation_comp != nullptr)
+		//{
+		//	AnimController* anim = nullptr;
+		//	if (animation_comp->GetAnimController() != nullptr)
+		//	{
+		//		anim = animation_comp->GetAnimController();
+		//	}
+		//	if (ImGui::IsMouseClicked(0) && ImGui::IsItemHovered() || animation_comp->playing == true)
+		//	{
+		//		
+		//		for (auto & playing_clip : anim->playing_clips)
+		//		{
+		//			if (playing_clip.playing == true && playing_clip.clip->name == node->state->clip->name)
+		//			{
+		//				ax::NodeEditor::SelectNode(node->id);
+		//			}
+		//		}
+		//	}
 
+		//	//for each transition currently on progress, animate the corresponding transition link
+		//	if (anim->active_transition && anim->apply_transition)
+		//	{
+		//		for (auto& link : links)
+		//		{
+		//			if (anim->active_transition->source_hash == link->transition->source_hash && anim->active_transition->target_hash == link->transition->target_hash)
+		//			{
+		//				ax::NodeEditor::Flow(link->id);
+		//			}
+		//			
+		//		}
+		//	}
+		//}
+
+		//for each animation clip that is currently playing, display circle on correspodning node->state
```

Comment:
If this code is not useful anymore better to deleted to keep the repo clean

Reply:
This code is the task regarding the Debug mode of the Node Editor. It threw some errors in the last testings so I commented it, in order to further debug it. We can erase it for now in develop, but I need it for debugging. Anabel pointed out to me to use a different already existing method.

Reply:
Cool then 😄